core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -178,6 +178,9 @@ private[spark] class Executor(
     /** Whether this task has been killed. */
     @volatile private var killed = false
 
+    /** Whether this task has been finished. */
+    @volatile private var finished = false
```

Comment:
You don't need to add `@volatile` since it's only accessed in the `synchronized` blocks. Could you also add `GuardedBy` like this:

```
    @GuardedBy("TaskRunner.this")
    private var finished = false
```